src/index.ts

```diff
@@ -447,14 +448,38 @@ class WorkerInfo extends AsynchronouslyCreatedResource {
     )
   }
 
-  async destroy(): Promise<void> {
-    await this.worker.terminate()
-    this.port.close()
-    this.clearIdleTimeout()
-    for (const taskInfo of this.taskInfos.values()) {
-      taskInfo.done(Errors.ThreadTermination())
-    }
-    this.taskInfos.clear()
+  async destroy(timeout?: number): Promise<void> {
+    let resolve: () => void
+    let reject: (err: Error) => void
+
+    const ret = new Promise<void>((res, rej) => {
+      resolve = res
+      reject = rej
+    })
+
+    const timer = timeout
+      ? setTimeout(
+          () => reject(new Error('Failed to terminate worker')),
+          timeout
+        )
+      : null
+
+    this.worker.terminate().then(() => {
```

Comment:
can't we just await the terminate instead of using the then keyword? if not, why?

Reply:
We cannot `await` as the outer Promise has to be the return value.

This outer Promise is required since `setTimeout` is used to decide when `destroy` function should reject. This would not work:

```js
async function destroy() {
  setTimeout(() => {
    throw new Error("Timeout error")
  }, 1000);

  await sleep(2000);
}
```

```js
destroy().then(() => console.log("success")).catch(() => console.log("Failed"))

> Uncaught Error: Timeout error
> success
```

Wrapping everything in a outer Promise helps:

```js
async function destroy() {
  let resolve, reject;
  const outerPromise = new Promise((res, rej) => {
    resolve = res
    reject = rej
  })
  
  setTimeout(() => {
    reject(new Error("Timeout error"))
  }, 1000);

  sleep(2000).then(resolve);

  return outerPromise;
}
```

```js
destroy().then(() => console.log("success")).catch(() => console.log("Failed"))

> Failed
```

Similar pattern is used in `runTask`:

https://github.com/tinylibs/tinypool/blob/d5e573865d7a50fcf40166da8d22cbc2f051a095/src/index.ts#L832-L838